src/fheroes2/world/world_loadmap.cpp

```diff
@@ -692,56 +694,93 @@ bool World::loadResurrectionMap( const std::string & filename )
     // Read and populate objects.
     const auto & townObjects = Maps::getObjectsByGroup( Maps::ObjectGroup::KINGDOM_TOWNS );
     const auto & heroObjects = Maps::getObjectsByGroup( Maps::ObjectGroup::KINGDOM_HEROES );
+    const auto & miscellaneousObjects = Maps::getObjectsByGroup( Maps::ObjectGroup::ADVENTURE_MISCELLANEOUS );
+    const auto & waterObjects = Maps::getObjectsByGroup( Maps::ObjectGroup::ADVENTURE_WATER );
+
+#if defined( WITH_DEBUG )
+    std::set<uint32_t> standardMetadataUIDs;
+    std::set<uint32_t> castleMetadataUIDs;
+    std::set<uint32_t> heroMetadataUIDs;
+    std::set<uint32_t> sphinxMetadataUIDs;
+    std::set<uint32_t> signMetadataUIDs;
+    std::set<uint32_t> adventureMapEventMetadataUIDs;
+#endif
 
     for ( size_t tileId = 0; tileId < map.tiles.size(); ++tileId ) {
         const auto & tile = map.tiles[tileId];
 
         for ( const auto & object : tile.objects ) {
             if ( object.group == Maps::ObjectGroup::KINGDOM_TOWNS ) {
-                const int race = ( 1 << townObjects[object.index].metadata[0] );
+#if defined( WITH_DEBUG )
+                castleMetadataUIDs.emplace( object.id );
+#endif
 
-                assert( race != Race::RAND );
+                const int color = ( 1 << Maps::getTownColorIndex( map, tileId, object.id ) );
+
+                int race = ( 1 << townObjects[object.index].metadata[0] );
+                const bool isRandom = ( race == Race::RAND );
+
+                if ( isRandom ) {
+                    assert( townObjects[object.index].objectType == MP2::OBJ_RANDOM_CASTLE || townObjects[object.index].objectType == MP2::OBJ_RANDOM_TOWN );
+
+                    const Kingdom & kingdom = GetKingdom( color );
+                    race = static_cast<uint8_t>( kingdom.GetRace() );
```

Comment:
For the neutral player we should pick a really random race. :)
```suggestion
                    race = ( color == Color::NONE ) ? Race::Rand() : static_cast<uint8_t>( kingdom.GetRace() );
```